File: FirstApp/FormView/KittyView.vb
Public Class KittyView

    Protected _currentCustomer As Customer

    'ENUM Dgv-> KittyView
    Public Enum DgvEnum
        CheckBoxColumn = 0
        SrNoColumn = 1
        KittyNoColumn = 2
        KittyTypeColumn = 3
        PendingInstalments = 4
        GivenAmountColumn = 5
        TotalAmountColumn = 6
        StatusColumn = 7
        DetailsButtonColumn = 8
        MakeEntryColumn = 9
        KittyIDColumn = 10
    End Enum

    'ENUM MessageDGv-> KittyView
    Public Enum MessageDgvEnum
        CheckBoxColumn = 0
        SrNoColumn = 1
        KittyNoColumn = 2
        KittyTypeColumn = 3
        PendingInstalments = 4
        GivenAmountColumn = 5
        TotalAmountColumn = 6
        StatusColumn = 7
        MessageBoxColumn = 8
        KittyIDColumn = 9
    End Enum

    Public Event ShowDetailsClicked(_kitty As Kitty)

    'FUNCTION:- Loads Kitty Data On To A DataGridView
    Public Sub LoadKittyData(Dgv As DataGridView)
        Dgv.Rows.Clear()
        Dim LastInstalmentList = ""
        Try
            'Looping Through All Owned Kitties
            For Each _kitty In _currentCustomer.OwnedKitties.OrderBasedOnStatus()

                Dgv.Rows.Add()
                Dim RowNo As Integer = Dgv.Rows.Count - 1

                'Adding Corresponding Values
                With Dgv.Rows(RowNo)
                    .Cells(DgvEnum.KittyNoColumn).Value = If(_kitty.KittyNo > 0, _kitty.KittyNo, "")
                    .Cells(DgvEnum.KittyTypeColumn).Value = _kitty.KittyType.ToCurrency
                    .Cells(DgvEnum.GivenAmountColumn).Value = _kitty.GivenAmount.ToCurrency
                    .Cells(DgvEnum.TotalAmountColumn).Value = $"{_kitty.TotalAmount.ToCurrency} + {_kitty.KittyInterest.ToCurrency}"
                    .Cells(DgvEnum.PendingInstalments).Value = _kitty.PendingInstalmentsInWords(IncludeInstalmentWord:=True)
                    .Cells(DgvEnum.StatusColumn).Value = _kitty.Status
                    If Dgv Is DgvMessage Then
                        .Cells(MessageDgvEnum.KittyIDColumn).Value = _kitty
                    Else
                        .Cells(DgvEnum.KittyIDColumn).Value = _kitty
                    End If
                End With

                With Dgv.Rows(RowNo).Cells(DgvEnum.PendingInstalments).Style
                    If _kitty.GetInstalmentsPending = -1 Or _kitty.GetInstalmentsPending >= 1 Then
                        .ForeColor = Color.Red
                    Else
                        .ForeColor = Color.Green
                    End If
                End With

                If _kitty.IsMatured Or _kitty.IsAvailed Or _kitty.IsInactive Then
                    With Dgv.Rows(RowNo)
                        .Cells(DgvEnum.CheckBoxColumn).Value = True
                        .Cells(DgvEnum.PendingInstalments).Value = String.Empty
                        .Cells(DgvEnum.CheckBoxColumn).Style.BackColor = SystemColors.Highlight
                        .Cells(DgvEnum.CheckBoxColumn).Style.ForeColor = SystemColors.Highlight
                        .Cells(DgvEnum.GivenAmountColumn).Style.BackColor = SystemColors.Highlight
                        .Cells(DgvEnum.TotalAmountColumn).Style.BackColor = SystemColors.Highlight
                        .Cells(DgvEnum.GivenAmountColumn).Style.ForeColor = Color.White
                        .Cells(DgvEnum.TotalAmountColumn).Style.ForeColor = Color.White
                        If _kitty.IsInactive Then .Cells(DgvEnum.StatusColumn).Style.ForeColor = Color.Silver
                        If _kitty.IsMatured Then .Cells(DgvEnum.StatusColumn).Style.ForeColor = Color.Gold
                        If _kitty.IsAvailed Then .Cells(DgvEnum.StatusColumn).Style.ForeColor = Color.Red
                    End With
                End If

            Next
        Catch ex As Exception
            MessageBox.Show($"[Kittyview] LoadKittyData : {ex.Message}")
        End Try

        SrNoUpdate(Dgv, DgvEnum.SrNoColumn)
    End Sub

    Public Sub KittyView_Load() Handles MyBase.Load
        _currentCustomer = Tag
        LoadKittyData(DgvMain)
    End Sub

    'EVENT:- ShowDetailsBUTTON MakeEntryBUTTON CLICK->(Shows KittyForm / UpdatesKittyID / ReLoadsKittyData)
    Public Sub Dgv_CellClick(sender As Object, e As DataGridViewCellEventArgs) Handles DgvMain.CellClick
        If e.RowIndex < 0 OrElse e.ColumnIndex < DgvEnum.DetailsButtonColumn Then Exit Sub

        Dim _kitty As Kitty = DgvMain.Rows(e.RowIndex).Cells(DgvEnum.KittyIDColumn).Value

        If e.ColumnIndex = DgvEnum.DetailsButtonColumn Then         'EVENT-> (Show Details) Click On Dvg

            'Invoke Event To Reload KittyView Data Because Changes Are Made TO Kitty
            RaiseEvent ShowDetailsClicked(_kitty)

        ElseIf e.ColumnIndex = DgvEnum.MakeEntryColumn Then         'EVENT-> Make Entry Click On Dvg

            If _kitty.IsMatured Then 'Raising Error:- When Making Entry On Matured Kitty's
                MessageBox.Show("[KittyView] This Kitty Is Matured, Can't Make Entry.", "EntryError")
                Exit Sub
            End If

            If _kitty.IsAvailed Then
                MessageBox.Show("[KittyView] This Kitty Is Cracked UnCrack To Continue Making Further Payments.", "Entry Error", MessageBoxButtons.OK, MessageBoxIcon.Hand)
                Exit Sub
            End If

            If MessageBox.Show($"Entry Of Amount {_kitty.KittyType.ToCurrency} To {_kitty.Owner.FullName("-")}", "Confirm", MessageBoxButtons.YesNo, MessageBoxIcon.Question) = DialogResult.No Then Exit Sub

            _kitty.AddRecord(Today.Date.ToShortDateString + _kitty.KittyType.ToString)
            _kitty.Save()
            LoadKittyData(DgvMain)
        End If
    End Sub

    'EVENT:- Message CheckBox Checked->(Shows Message Dgv) 
    Private Sub MarriedCheckBox_CheckedChanged(sender As CheckBox, e As EventArgs) _
        Handles MessageCheckBox.CheckedChanged
        If sender.Checked Then
            LoadKittyData(DgvMessage)
            DgvMain.Visible = False
            DgvMessage.Visible = True
        Else
            DgvMain.Visible = True
            DgvMessage.Visible = False
            ActivateButton(Activate:=False)
        End If
    End Sub

    'FUNCTION:- Activate Send Message Button
    Private Sub ActivateButton(Activate As Boolean)
        If Activate Then
            SendMessageButton.ForeColor = Color.MediumPurple
            SendMessageButton.Enabled = True
        Else
            SendMessageButton.ForeColor = Color.DimGray
            SendMessageButton.Enabled = False
        End If
    End Sub

    'EVENT:- CheckBox On Message Dgv Clicked->(Checks CheckBox / Activates Button)
    Private Sub DgvMessage_CellClick(sender As DataGridView, e As DataGridViewCellEventArgs) Handles DgvMessage.CellClick
        If e.ColumnIndex = MessageDgvEnum.MessageBoxColumn And e.RowIndex >= 0 Then
            With sender.Rows(e.RowIndex).Cells(e.ColumnIndex)
                If .Value Then
                    .Value = False
                Else
                    .Value = True
                End If
            End With

            For i = 0 To sender.Rows.Count - 1
                If sender.Rows(i).Cells(DgvEnum.DetailsButtonColumn).Value Then
                    ActivateButton(Activate:=True)
                    Exit Sub
                End If
            Next
            ActivateButton(Activate:=False)
        End If
    End Sub

    Private Sub SendMessageButton_Click(sender As Object, e As EventArgs) Handles SendMessageButton.Click
        Dim SelectedKitty As New List(Of Kitty)
        For RowNo = 0 To DgvMessage.Rows.Count - 1
            If DgvMessage.Rows(RowNo).Cells(MessageDgvEnum.MessageBoxColumn).Value Then
                SelectedKitty.Add(DgvMessage.Rows(RowNo).Cells(MessageDgvEnum.KittyIDColumn).Value)
            End If
        Next

        Dim Fm As New Form
        Fm = Messenger
        Fm.Tag = SelectedKitty
        Fm.Show()
    End Sub

End Class

File: FirstApp/SharedForms/KittyView.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;
using Microsoft.VisualBasic.CompilerServices;

namespace FirstApp
{
    public partial class KittyView
    {

        protected Customer _currentCustomer;

        // ENUM Dgv-> KittyView
        public enum DgvEnum
        {
            CheckBoxColumn = 0,
            SrNoColumn = 1,
            KittyNoColumn = 2,
            KittyTypeColumn = 3,
            PendingInstalments = 4,
            GivenAmountColumn = 5,
            TotalAmountColumn = 6,
            StatusColumn = 7,
            DetailsButtonColumn = 8,
            MakeEntryColumn = 9,
            KittyIDColumn = 10
        }

        // ENUM MessageDGv-> KittyView
        public enum MessageDgvEnum
        {
            CheckBoxColumn = 0,
            SrNoColumn = 1,
            KittyNoColumn = 2,
            KittyTypeColumn = 3,
            PendingInstalments = 4,
            GivenAmountColumn = 5,
            TotalAmountColumn = 6,
            StatusColumn = 7,
            MessageBoxColumn = 8,
            KittyIDColumn = 9
        }

        public event ShowDetailsClickedEventHandler ShowDetailsClicked;

        public delegate void ShowDetailsClickedEventHandler(Kitty _kitty);

        public KittyView()
        {
            InitializeComponent();
        }

        // FUNCTION:- Loads Kitty Data On To A DataGridView
        public void LoadKittyData(DataGridView Dgv)
        {
            Dgv.Rows.Clear();
            string LastInstalmentList = "";
            try
            {
                // Looping Through All Owned Kitties
                foreach (var _kitty in _currentCustomer.OwnedKitties().OrderBasedOnStatus())
                {

                    Dgv.Rows.Add();
                    int RowNo = Dgv.Rows.Count - 1;

                    // Adding Corresponding Values
                    {
                        var withBlock = Dgv.Rows[RowNo];
                        withBlock.Cells[(int)DgvEnum.KittyNoColumn].Value = _kitty.KittyNo > 0 ? _kitty.KittyNo : "";
                        withBlock.Cells[(int)DgvEnum.KittyTypeColumn].Value = _kitty.KittyType.ToCurrency();
                        withBlock.Cells[(int)DgvEnum.GivenAmountColumn].Value = _kitty.GivenAmount().ToCurrency();
                        withBlock.Cells[(int)DgvEnum.TotalAmountColumn].Value = $"{_kitty.TotalAmount().ToCurrency()} + {_kitty.KittyInterest.ToCurrency()}";
                        withBlock.Cells[(int)DgvEnum.PendingInstalments].Value = _kitty.PendingInstalmentsInWords(IncludeInstalmentWord: true);
                        withBlock.Cells[(int)DgvEnum.StatusColumn].Value = _kitty.Status();
                        if (ReferenceEquals(Dgv, DgvMessage))
                        {
                            withBlock.Cells[(int)MessageDgvEnum.KittyIDColumn].Value = _kitty;
                        }
                        else
                        {
                            withBlock.Cells[(int)DgvEnum.KittyIDColumn].Value = _kitty;
                        }
                    }

                    {
                        var withBlock1 = Dgv.Rows[RowNo].Cells[(int)DgvEnum.PendingInstalments].Style;
                        if (_kitty.GetInstalmentsPending() == -1 | _kitty.GetInstalmentsPending() >= 1)
                        {
                            withBlock1.ForeColor = Color.Red;
                        }
                        else
                        {
                            withBlock1.ForeColor = Color.Green;
                        }
                    }

                    if (_kitty.IsMatured() | _kitty.IsAvailed | _kitty.IsInactive)
                    {
                        {
                            var withBlock2 = Dgv.Rows[RowNo];
                            withBlock2.Cells[(int)DgvEnum.CheckBoxColumn].Value = true;
                            withBlock2.Cells[(int)DgvEnum.PendingInstalments].Value = string.Empty;
                            withBlock2.Cells[(int)DgvEnum.CheckBoxColumn].Style.BackColor = SystemColors.Highlight;
                            withBlock2.Cells[(int)DgvEnum.CheckBoxColumn].Style.ForeColor = SystemColors.Highlight;
                            withBlock2.Cells[(int)DgvEnum.GivenAmountColumn].Style.BackColor = SystemColors.Highlight;
                            withBlock2.Cells[(int)DgvEnum.TotalAmountColumn].Style.BackColor = SystemColors.Highlight;
                            withBlock2.Cells[(int)DgvEnum.GivenAmountColumn].Style.ForeColor = Color.White;
                            withBlock2.Cells[(int)DgvEnum.TotalAmountColumn].Style.ForeColor = Color.White;
                            if (_kitty.IsInactive)
                                withBlock2.Cells[(int)DgvEnum.StatusColumn].Style.ForeColor = Color.Silver;
                            if (_kitty.IsMatured())
                                withBlock2.Cells[(int)DgvEnum.StatusColumn].Style.ForeColor = Color.Gold;
                            if (_kitty.IsAvailed)
                                withBlock2.Cells[(int)DgvEnum.StatusColumn].Style.ForeColor = Color.Red;
                        }
                    }

                }
            }
            catch (Exception ex)
            {
                MessageBox.Show($"[Kittyview] LoadKittyData : {ex.Message}");
            }

            ProjectModule.SrNoUpdate(Dgv, (int)DgvEnum.SrNoColumn);
        }

        public void KittyView_Load()
        {
            _currentCustomer = (Customer)Tag;
            LoadKittyData(DgvMain);
        }

        // EVENT:- ShowDetailsBUTTON MakeEntryBUTTON CLICK->(Shows KittyForm / UpdatesKittyID / ReLoadsKittyData)
        public void Dgv_CellClick(object sender, DataGridViewCellEventArgs e)
        {
            if (e.RowIndex < 0 || e.ColumnIndex < (int)DgvEnum.DetailsButtonColumn)
                return;

            Kitty _kitty = (Kitty)DgvMain.Rows[e.RowIndex].Cells[(int)DgvEnum.KittyIDColumn].Value;

            if (e.ColumnIndex == (int)DgvEnum.DetailsButtonColumn)         // EVENT-> (Show Details) Click On Dvg
            {

                // Invoke Event To Reload KittyView Data Because Changes Are Made TO Kitty
                ShowDetailsClicked?.Invoke(_kitty);
            }

            else if (e.ColumnIndex == (int)DgvEnum.MakeEntryColumn)         // EVENT-> Make Entry Click On Dvg
            {

                if (_kitty.IsMatured()) // Raising Error:- When Making Entry On Matured Kitty's
                {
                    MessageBox.Show("[KittyView] This Kitty Is Matured, Can't Make Entry.", "EntryError");
                    return;
                }

                if (_kitty.IsAvailed)
                {
                    MessageBox.Show("[KittyView] This Kitty Is Cracked UnCrack To Continue Making Further Payments.", "Entry Error", MessageBoxButtons.OK, MessageBoxIcon.Hand);
                    return;
                }

                if (MessageBox.Show($"Entry Of Amount {_kitty.KittyType.ToCurrency()} To {_kitty.Owner.FullName("-")}", "Confirm", MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.No)
                    return;

                _kitty.AddRecord(DateTime.Today.Date.ToShortDateString() + _kitty.KittyType.ToString());
                _kitty.Save();
                LoadKittyData(DgvMain);
            }
        }

        // EVENT:- Message CheckBox Checked->(Shows Message Dgv) 
        private void MarriedCheckBox_CheckedChanged(CheckBox sender, EventArgs e)
        {
            if (sender.Checked)
            {
                LoadKittyData(DgvMessage);
                DgvMain.Visible = false;
                DgvMessage.Visible = true;
            }
            else
            {
                DgvMain.Visible = true;
                DgvMessage.Visible = false;
                ActivateButton(Activate: false);
            }
        }

        // FUNCTION:- Activate Send Message Button
        private void ActivateButton(bool Activate)
        {
            if (Activate)
            {
                SendMessageButton.ForeColor = Color.MediumPurple;
                SendMessageButton.Enabled = true;
            }
            else
            {
                SendMessageButton.ForeColor = Color.DimGray;
                SendMessageButton.Enabled = false;
            }
        }

        // EVENT:- CheckBox On Message Dgv Clicked->(Checks CheckBox / Activates Button)
        private void DgvMessage_CellClick(DataGridView sender, DataGridViewCellEventArgs e)
        {
            if (e.ColumnIndex == (int)MessageDgvEnum.MessageBoxColumn & e.RowIndex >= 0)
            {
                {
                    var withBlock = sender.Rows[e.RowIndex].Cells[e.ColumnIndex];
                    if (Conversions.ToBoolean(withBlock.Value))
                    {
                        withBlock.Value = false;
                    }
                    else
                    {
                        withBlock.Value = true;
                    }
                }

                for (int i = 0, loopTo = sender.Rows.Count - 1; i <= loopTo; i++)
                {
                    if (Conversions.ToBoolean(sender.Rows[i].Cells[(int)DgvEnum.DetailsButtonColumn].Value))
                    {
                        ActivateButton(Activate: true);
                        return;
                    }
                }
                ActivateButton(Activate: false);
            }
        }

        private void SendMessageButton_Click(object sender, EventArgs e)
        {
            var SelectedKitties = new List<Kitty>();
            for (int RowNo = 0, loopTo = DgvMessage.Rows.Count - 1; RowNo <= loopTo; RowNo++)
            {
                if (Conversions.ToBoolean(DgvMessage.Rows[RowNo].Cells[(int)MessageDgvEnum.MessageBoxColumn].Value))
                {
                    SelectedKitties.Add((Kitty)DgvMessage.Rows[RowNo].Cells[(int)MessageDgvEnum.KittyIDColumn].Value);
                }
            }

            var Fm = new MessengerForIndividual() { SelectedKitties = SelectedKitties };

            Fm.Show();
        }

        private void KittyView_Load(object sender, EventArgs e) => KittyView_Load();

    }
}